cmd/bootstrap/cmd/seal.go

```diff
@@ -78,3 +80,18 @@ func constructRootResultAndSeal(
 
 	return result, seal
 }
+
+// rootEpochTargetEndTime computes the target end time for the given epoch, using the given config.
+// CAUTION: the variables `flagEpochTimingRefCounter`, `flagEpochTimingDuration`, and
+// `flagEpochTimingRefTimestamp` must contain proper values. You can either specify a value for
+// each config parameter or use the function `validateOrPopulateEpochTimingConfig()` to populate the varaibles
+// from defaults.
+func rootEpochTargetEndTime() uint64 {
+	if flagEpochTimingRefTimestamp == 0 || flagEpochTimingDuration == 0 {
+		panic("invalid epoch timing config: must specify ALL of --epoch-target-end-time-ref-counter, --epoch-target-end-time-ref-timestamp, and --epoch-target-end-time-duration")
+	}
+	if flagEpochCounter < flagEpochTimingRefCounter {
+		panic("invalid epoch timing config: reference epoch counter must be before root epoch counter")
+	}
+	return flagEpochTimingRefTimestamp + (flagEpochCounter-flagEpochTimingRefCounter)*flagEpochTimingDuration
+}
```

Comment:
would it make sense to include a sanity check here that the root Epoch's Target End Time is in the future from now? I don't think the code is equipped to handle this edge case. And I assume that accidental configuration errors might happen. After all, the parameters are set by humans.

Reply:
Added in https://github.com/onflow/flow-go/pull/5071/commits/c3c6fc9509ce79e1257697134e2656b8efe62464